test_plans/math_array.asciidoc

```diff
@@ -0,0 +1,449 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for math array
+
+This is a test plan for the math array types described in SYCL 2020 Section 4.14.3.
+
+Estimated development time is four days.
+
+== Testing scope
+
+This plan does not include negative tests.
+
+=== Backend coverage
+
+All the tests described below are performed for any SYCL implementation.
+
+=== Device coverage
+
+All the tests described below are performed only on the default device that is selected on the CTS command line.
+
+=== Type coverage
+
+The interface tests and memory layout and alignment tests (Sections 2.1 and 2.3 below) are performed using each of the following types as the underlying type the `DataT` template variable of `marray`, unless stated otherwise.
+
+In regular mode:
+
+* `char`
+* `int`
+* `float`
+* `std::byte`
+* `std::int8_t`
+* `std::int32_t`
+* A user-defined struct that meets the NumericType requirements of the C++ standard.
+
+In addition, if the device has `aspect::fp64`:
+
+* `double`
+
+In addition, if the device has `aspect::fp16`:
+
+* `half`
+
+In full conformance mode:
+
+* `signed char`
+* `unsigned char`
+* `short`
+* `unsigned short`
+* `unsigned int`
+* `long`
+* `unsigned long`
+* `long long`
+* `unsigned long long`
+* `bool`
+* `std::uint8_t`
+* `std::int16_t`
+* `std::uint16_t`
+* `std::uint32_t`
+* `std::int64_t`
+* `std::uint64_t`
+
+=== Dimension coverage
+
+The interface tests and memory layout and alignment tests (Sections 2.1 and 2.3 below) are performed using each of the following dimensions for the `NumElements` template variable of `marray`: `1`, `2`, `4`, `8`, `16`, `32`, and `64`, unless stated otherwise.
```

Comment:
The `NumElements` template parameter is allowed to be any value of type `size_t`, even ones that are not a power of two.  I think you should check at least one non-power-of-two value (e.g. `5` might be good).  Technically, I think `0` is also allowed.  Would that fall out for free when you write the test, or will it require special handling?  If it falls out for free, it's worth testing.  If `0` requires a lot of extra work, we can save it until after we get the basic tests written.

Reply:
No, 0 is not allowable, see 4.14.3.

Reply:
Thanks for the suggestion, I have added a non-power-of-two value.

I agree with Melirius, the spec states:
> The number of elements parameter, `NumElements`, is a positive value of the `std::size_t` type.

In the case that we are interpreting it incorrectly, `0` can be added without much effort (some tests should slightly be modified when `NumElements` is `0`, such as the iterator test).